### ISSUE TITLE
No Command To Run External Files

### ISSUE THREAD

**Issue by tbpassin** [leo-editor/leo-editor#3035]
Leo does not seem to have a command to run an external file (I mean an @file, @clean, etc. external file).  There are various class methods that could be adapted or used in such a command but I haven't found a minibuffer command to do so. I have my own command, and I imagine many other people do too, but it's not really production grade at this point.  

CTRL-B doesn't always do the job.  For example, it can't run some Tk external programs because Leo is running Qt and they don't play together:

```
ImportError: Cannot load backend 'TkAgg' which requires the 'tk' interactive framework, as 'qt' is currently running
```

Yes, it's always possible to create a node that can run a specific file using, for example, subprocess.run() or os.system(), but that's clumsy and time-consuming compared with running a minibuffer command.

I suggest that Leo should have a minibuffer command to run an external file.  Most all IDEs do.  At a minimum, the command should be able to run Python programs but preferably any program for which the user has a runtime executable (my own command uses an ini file to specify what executable to run for which kind of program (more Leonine would be a setting and an `@data` node).  The command should work on Windows, Linux, and Mac, and for both console and GUI programs.  Running the external program should not block Leo.

**Comment by edreamleo** [comment]
@tbpassin Thanks for these comments. This command makes sense and doesn't seem to be a security problem.

Are you willing to give it a go? Here's what I would do:

- Find the corresponding `@<file>` node using p.isAnyAtFileNode().
  Start with c.p and look up the tree, looking for the root node.
  It probably doesn't make sense to look down the tree.
- Punt (warn) if no `@<file>` node can be found.
- Execute `g.executeFile(root)`.

**Comment by tbpassin** [comment]
Yes, I will take a run at it.  I need to study `g.executeFile()` to see if it needs more information to know what program to use (e.g., julia 1.6 vs 1.8 for .jl files, etc.)

**Comment by tbpassin** [comment]
It looks like `g.executeFile()` blocks, and also only works for Python files, so it's probably only part of the solution.

One thing to settle is whether stdout/stderr should be displayed inside Leo or not.  I've always been happy with seeing them on the external terminal that opened to launch the external file, and that would be easier, but most IDEs can display them on a panel inside the IDE.  OTOH, capturing them without blocking Leo would probably require using threads, and so be more complicated.

Perhaps displaying stdout/stderr in a new tab in the log frame would be a good solution.

**Comment by edreamleo** [comment]
@tbpassin Thanks for handling this issue.

**Comment by tbpassin** [comment]
I have a working prototype, that can launch external files in Linux and Windows, with these features:

1. It opens a persistent terminal;
2. All output displays on this terminal;
3. Leo is not blocked;
4. The external processor to use (if not python or shell) (e.g., lua, julia, etc.) can be specified (currently in a dictionary, eventually by an `@data` setting node, I think);
5. If a suitable processor is installed but not included in the processor definition dictionary, the file will still get run in Windows if cmd.exe knows how to run it by its extension.  In Linux, if the file includes a shebang line, the shell will run it (if it's not an `@clean` tree, the shebang line needs to start with `@first #! ....`).
6. The program to use depends on the file's language which is taken from its @language directive, or its file extension if there is no directive.

I think this is for the most part acceptable, but:

1. In Linux, the gnome-terminal gets invoked and I suppose that's not always available depending on the distro.  xterm is probably more widely available but on my virtual machines looks unappetizing compared ti gnome-terminal.

2. There is currently no way to pass command-line parameters.  It's possible to imagine various ways to specify them, but they would all take some effort on the part of the user.  It would be good to solve this in a good, Leonine way.  (Most of the times that I launch external programs with my own command, I haven't needed to pass parameters).

There is also the possibility of capturing program output and displaying it in Leo instead of on a persistent terminal.  I picture this output as displaying in a new tab in the log pane.  The tab would run a `communicate()` thread similar to the background process manager, and display when the process is finished.  

The problem to be solved here is that you don't always want to capture any output.  For example, if you run a GUI program, it will probably run for a long time and you won't want the tab to be waiting around for program termination.  So there would have to be some way to specify not to capture the output, and not to wait for the launched program to terminate.

My current thinking is to complete work on the current version (output displays outside of Leo, no command line parameters).  After that, we can decide if we want to extend the minibuffer command, or more likely, develop a new one with a new name with parameters and/or captured output.

Suggestions, please!

**Comment by edreamleo** [comment]
@tbpassin Many thanks for your thorough work. You understand the issue better than I. I trust your judgement. Have at it!

**Comment by SegundoBob** [comment]
Thomas,

Your ideas are interesting, but I don't have time right now to properly consider them.  The problem you are addressing is also partially addressed by the Leo-Editor plugin Leo-Babel (which I wrote) and more or less completely addressed by the emacs Babel extension.

SegundoBob###

**Comment by tbpassin** [comment]
The Leo-Babel subtree in LeoPyRef seems to be messed up.  The files are there in the directory tree on disk, but some of them are missing from LeoPyRef itself.  These include _babel.py_ and _babel_api.py_.

**Comment by edreamleo** [comment]
@tbpassin Feel free to create a PR for LeoPyRef. Thanks.

**Comment by tbpassin** [comment]
This is turning out to be a real bear, because different Linux distros do things differently and it's hard to discover how to make it all work.  You need to open a terminal-emulator and also a shell, and to have that terminal stay open after your external program finishes so you can read the output.  And you want that to happen without blocking Leo.

I have devised a way to discover which terminal-emulator is in use.  I can use the same command with both Gnome-based and xfce-based window managers, and have everything happen as intended without blocking.  The commands are a little different for KDE-based systems like KUbuntu and OpenSuse, and opening the external program works, but it blocks Leo.

In general, I have read that almost all newer Linux distros have the bash shell, so that simplifies things.  I can discover which shell is running if it is not bash, but there is no way to discover the right command arguments to use, so there would have to be a way to special-case them, which can be done but they would need to be known.  

Similarly for the terminal-emulator, it is claimed that "most" modern distros have xterm, but that turns out not to be true.  Gnome-based systems have gnome-terminal, xfce-based systems have xfce4-terminal, and KDE-based systems have konsole.

I have not found a breakdown of which terminal emulator comes standard with which distro.  But I can't create a new VM for each one, since there are so many.  I'm fairly sure that the calling args to use can be stored in an `@data` node; the problem is how to populate it.

Another wrinkle is to know when you can use a shebang line to run a program and when you need to specify a processor, like, say, Julia or Lua. To know how to run a program with an arbitrary processor (if it does not have a shebang line) we will probably need to include that information in another `@data` node.

MacOS is another story.  I think the information is out there, and all Macs will probably work the same (unlike the case for Linux), but I don't have a MacOS system to test with.

I'm not giving up, but this effort will take some time and the solution may not work on all Linux distros.  At least it already works on gnome-based system and xfce-based (including Arch-linux derived) systems, and those are probably in the majority.

**Comment by tbpassin** [comment]
Progress!  I can now get the execute flag or argument from the terminal's help message, and I can also get the name of the terminal in use.  With the execute flag, I can open a shell in a new terminal and run the external program with the shell.

What I can't do so far is to make sure that opening that new terminal will be non-blocking.  It is for some (gnome-shell) and not others.  I don't know if this is distro-dependent or terminal-dependent.

But progress is good...

**Comment by tbpassin** [comment]
All right, the problem of keeping the terminal open after our program closes is solved by the Popen argument `start_new_session=True`.  I had thought that it was a Windows-only argument, but that was wrong.

The next thing to tackle is how to know what program to use to launch our external program if it is not a Python program and does not have a shebang line.

**Comment by tbpassin** [comment]
A status report.  I have a minibuffer/command that seems to work well on both Windows and multiple flavors of Linux.  For Linux, the challenge was to be able to launch a terminal that would run the command and then stay open after the command ended.

It was also tricky to work out how to have the newly opened terminal launch a GUI program and stay open.   To get the terminal to stay open, I ended up telling the shell to read from input after it launches our external program.  So it stays there waiting for someone to type something.  I couldn't get anything else to work across all six Linux distros I have to work with.

This turned out to require knowing both the shell (bash seems to be almost always available) and the terminal.  From the terminal Help screen we have to parse out the right option to use to run a shell and have it execute a command; it's not the same for all terminals.   Well, either that or hard-code the option for known terminals.  I'm trying to parse it out, but maybe I should also hard-code the known ones.  Trouble is, the options can change: I had at least one terminal warn me about a possible upcoming change.

After that, we need to be able to keep the shell open after the command returns.  That's so we can see any output or error messages.  And remember also that the external program to be run might be a GUI program.

We also have to figure out what program to use to run our external file.  Turns out the xdg-* method - frequently recommended - is not reliable.  So I check if there is a shebang line and use that.  If not, I use the file extension, and consult a lookup table to get the program (e.g., ruby, julia, etc.).  The lookup table can be overridden and extended by an `@data` node.  The full path to the program can be specified in that data node if it's not on the PATH - which happened to me at least once.

This command is working well, as I said, and I've wrung out a number of bugs to the point I think it's basically ready for prime time.  One thing is to decide where it should live in _LeoPyRef_.
